runtime/compare.c

```diff
@@ -99,18 +99,17 @@ static intnat compare_val(value v1, value v2, int total)
 static void run_pending_actions(struct compare_stack* stk,
                                 struct compare_item* sp)
 {
-  value exn;
+  caml_result result;
   value* roots_start = (value*)(stk->stack);
   size_t roots_length =
     (sp - stk->stack)
     * sizeof(struct compare_item) / sizeof(value);
   Begin_roots_block(roots_start, roots_length);
-  exn = caml_do_pending_actions_exn();
+  result = caml_do_pending_actions_result();
   End_roots();
-  if (Is_exception_result(exn)) {
-    exn = Extract_exception(exn);
+  if (result.is_exception) {
     compare_free_stack(stk);
-    caml_raise(exn);
+    caml_raise(result.data);
```

Comment:
I tried to think about style that feel less like is violates the abstraction of the caml_result type. `caml_run_result` (with a better name) would be better, I think; or an `exn = result.data` following immediately the test for `is_exception`, imitating the style used previously. This is only important here insofar as we want to think about which style one would recommend to programmers for resource clean-up.

Reply:
I went for the "binding immediately after the test" as a form of pattern-matching:

```c
  if (result.is_exception) {
    value exn = result.data;
    compare_free_stack(stk);
    caml_raise(exn);
  }
```